keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    // ObservedGeneration is used by the Application Controller to report the last Generation of an Application 
+    // that it has observed.
+    ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"int64,1,opt,name=observedGeneration"`
+    
+    // Installed is used by the Application Controller ot report the installed Components and Dependencies of  
+    // an Application.
+    Installed []string `json:"installed,omitempty" protobuf:"bytes,2,opt,name=installed"`
+    
+    // Ready is used to report the readiness of an Application. If an Application contains an ApplicationHealthCheck, 
+    // this field corresponds directly to the Readiness of the Pod generated from that health check. If the 
+    // Application does not contain an ApplicationHealthCheck, this field is set to Unknown.
+    Ready ConditionStatus json:`"ready,omitempty" protobuf:"bytes,3,opt,name=ready"`
+}
+
+// ApplicationHealthCheck contains a template for a Pod that performs an Application health check. The readiness 
+// of the Pod decides the readiness of the Application.
+type ApplicationHealthCheck struct {
+    
+    // Template is used to specify a Pod whose readiness will be used to determine the readiness of an Application.
+    Template v1.PodTemplateSpec
+}
+// ApplicationSpec is the specification object for an Application. It specifies the Application's components, 
+// dependencies, and an optional health check. Its selector can be used to select application components.
+type ApplicationSpec struct {
+    // Type is the type of the application (e.g. WordPress, MySQL, Cassandra).
+    Type string `json:"type" protobuf:"bytes,1,name=type"`
+    
+    // Components is a map of the applications components to the kinds created by the application (e.g. Pods, Services, 
+    // Deployments, CRDS, etc)
+    Components map [string]metav1.GroupVersionKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
+    
+    // HealthCheck is an optional application level health check.
+    HealthCheck* ApplicationHealthCheck `json:"healthCheck" protobuf:"bytes,5,opt,name=healthCheck"`
+    
+    // Version is an optional version indicator for the application.
+    Version string `json:"version,omitempty" protobuf:"bytes,6,opt,name=version"`
+    
+    // AboutURL is a URL pointing to any additional information about the application.
+    AboutURL string `json:"url,omitempty" protobuf:"bytes,7,opt,name=url"`
```

Comment:
this feels like it could be handled via some well defined annotations which would be more flexible/extensible..does it need to be a first class field?
(i think we're going to have a hard time agreeing on the right set of "application descriptor metadata" so i'd rather see some proposed annotations and let people easily extend/mutate that set of annotations for their tools/uses)

Reply:
(same comment on most of the other fields being proposed here).

Reply:
I feel the proposal does a good job selecting the minimal set of descriptor metadata. Fields like "version" and "description" at least have been proposed multiple times and probably do stand a chance of getting agreement. The benefit of having some first-class descriptor would be significant, especially UIs to display basic information about off-the-shelf apps (e.g. from Helm).

Reply:
Would suggest grouping descriptive fields (i.e. intended for human consumption) under some kind of "Descriptor" object

Reply:
Agree about grouping descriptor fields, disagree that "AboutURL" and "LicenseURL" and such qualify as minimal though.